### DISCUSSION TITLE
Advice/Guidance on bitmap manipulation in python/low-level driver

### DISCUSSION THREAD

**Discussion by DaveWK** [mathoudebine/turing-smart-screen-python#212] [General]
Hi,

Thanks for putting this out; these are a neat little devices that seem to have some good potential.

I am interested in the "low-level" functionality of the sdk and was hoping you could give me some tips.

My impression based on the doc is that pushing the entire bitmap for a screen is a time intensive process, so the best approach is to push the "full" bitmap to the device only when the program starts, then after the "init" to only push updates to sections of the bitmap at a time for performance.

For that reason, I think the best approach would be to make a python class that maintains the full state of the bitmap, and have helper functions capable of making a diff between the current bitmap and the desired bitmap, and then being able to push only the updated portions to the device rather than the whole bitmap.

I was wondering if you/anyone was aware of  any python libraries/packages capable of holding an entire bitmap in memory and producing a diff between two bitmaps? It seems like if I had this, then it would be trivial to make CRUD API's for exposing to other programs.

**Comment by mathoudebine** [comment]
Hi @DaveWK

You're right that the refresh of the complete screen is very time consuming. What you describe is what is done by this project and also by the original vendor apps: the background full bitmap is displayed once on program startup, then only parts of the display are being updated.

Unfortunately i don't know any Python libraries that would allow diff between bitmaps. You could maybe do it manually, if you can access the bitmaps like an array of pixels?
